### PR TITLE
unify udev rules, use ceph-disk trigger ...

### PR DESCRIPTION
**Pull request by liewegas**
We have problems with 2 sets of udev rules because on debian the init system
choice is a runtime option.  Instead, funnel all udev events to ceph-disk
trigger so we can Do The Right Thing.

Note that the dmcrypt paths need tob e carefully tested.  The old udev rule
did the cryptsetup, then a loop waiting for the resulting device to appear,
then an explicit activate.  Not sure if we need to do teh same here.

### CODE REVIEW

**Comment by liewegas** [comment]
@ddiss @branto1 Does this look reasonable?

File: src/ceph-disk [hunk 1]
```diff
@@ -2943,6 +2952,64 @@ def main_zap(args):
...
+                '/sbin/cryptsetup',
+                '--key-file',
+                '/etc/ceph/dmcrypt-keys/{id}.luks.key luksOpen {dev} {id}'.format(
+                    id=args.partid,
+                    dev=args.dev,
+            ]
+        )
+
+    else:
+        raise Error('unrecognized partition type %s', args.parttype)
```
Comment (dachary) [inline src/ceph-disk]: The current udev logic for dmcrypt looks more like:

* cryptsetup
* udevadm settle (well, it's sleep right now but maybe it should be that instead)
* ceph-disk activate

**Comment by dachary** [comment]
Modulo the cryptsetup remark it looks like a great simplification :-)

**Comment by dachary** [comment]
If you send that to the gibuilders I can try a run of the ceph-disk suite and see if things have improved.

File: udev/95-ceph-osd.rules [hunk 2]
```diff
@@ -1,53 +1,53 @@
-# activate ceph-tagged partitions
+# OSD_UUID
 ACTION=="add", SUBSYSTEM=="block", \
   ENV{DEVTYPE}=="partition", \
   ENV{ID_PART_ENTRY_TYPE}=="4fbd7e29-9d25-41b8-afd0-062c0ceff05d", \
-  RUN+="/usr/sbin/ceph-disk activate /dev/$name"
+  OWNER:="ceph", GROUP:="ceph", MODE:="660", \
+  RUN+="/usr/sbin/ceph-disk trigger activate /dev/$name $env{ID_PART_TYPE_UUID} $env{ID_PART_ENTRY_UUID}"
```
Comment (trhoden) [inline udev/95-ceph-osd.rules]: I think the `activate` flag to ceph-disk here is extraneous in all of these?

File: src/ceph-disk [hunk 3]
```diff
@@ -2604,6 +2613,22 @@ def get_dev_fs(dev):
     else:
         return None
 
+def get_dev_udev_properties(dev):
```
Comment (dachary) [inline src/ceph-disk]: get_partition_type and get_partition_uuid do the same

File: src/ceph-disk [hunk 4]
```diff
@@ -2943,6 +2968,184 @@ def main_zap(args):
...
+                'create',
+                partid,
+                args.dev,
+            ]
+        )
+        newdev='/dev/mapper/' + partid
+        count=0
+        while not os.path.exists(newdev) and count <= 10:
+            time.sleep(1)
+            count += 1
```
Comment (dachary) [inline src/ceph-disk]: raise error here as activate will fail anyway ?

**Comment by dachary** [comment]
two nits but otherwise looks great :-)

**Comment by branto1** [comment]
Apart from that one thing, lgtm.

**Comment by ddiss** [comment]
Some of the hard-coded binary paths may be problematic. Nevertheless, I think this looks like a nice clean-up.

**Comment by dachary** [comment]
Here is an idea for dmcrypt : when udev event on /dev/vdb1 is seen, crypsetup is called. When a udev event on /dev/dm-X is seen and it is a slave to /dev/vdb1 and /dev/vdb1 is a Ceph dmcrypt, it is activated. The /sys/.../dm/name file can be used to retrieve the name that matches the key. Ilya says we can rely on an udev event every time a devicemapper file is created. 

That's not necessarily simpler than the current logic though. The only difference is that it does not rely on the /dev/mapper/XXX name but on the /sys/.../dm/name which may be more reliable.

I'm not sure which solution is better.

**Comment by mbroz** [comment]
You can safely use /sys/*/dm/name (it is not very old old kernels though). Ypou can also use lsblk, which pareses /sys information for you. (Anyway /dev/mapper/* must be reliable, if not, code is doing something wrong.)

Anyway, while this patchset solves many problems, you can still see occasionally some problems:
- udev rules react on "add" event (this is the first event and is sent always only once unless someone triggers it artificially), But there are usually several "change" events following it - all these events causes udev to rescan the device (using blkid or other tools).
If the systemd job run in parallel, it can happen that some scan has block device open (and some mapping could fail because device is busy). If this happens, code probably should retry operation (or at least call udevadm settle before trying). Just FYI :)

**Comment by dachary** [comment]
discovered dmcrypt journal activation is broken, fixing it while testing this branch http://tracker.ceph.com/issues/12897

**Comment by dachary** [comment]
@mbroz interesting. So there is no way for ceph-disk to grab the device and not race with something else ?

**Comment by mbroz** [comment]
@dachary udev is designed to run asynchronously. so in principle you should retry if device is busy and you know it is "your" device... (blkid scan should be very quick but there can be custom rules installed). I would ignore it for now, in the worst case easy workaround should be to add "udevadm settle" before mapping.

Anyway, if there are problems later with dmcrypt/LUKS devices let me know. (The whole story of udev & device-mapper is problematic and boring... specifically in Debian which uses modified storage rules ;)

This PR should solve major problem with old Ceph udev rules design (running cryptsetup in udev directly), I didn't try it yet myself but just from reading code: LGTM...

**Comment by dachary** [comment]
Needs to cherry-pick https://github.com/ddiss/ceph/commit/82a84285e35bc7e7c4669388ae3a6b79e8d2ab0b to fix http://tracker.ceph.com/issues/12897

**Comment by dachary** [comment]
For dmcrypt the trigger implementation is not enough as it does not have the keydir nor the key length in the case of plain dmcrypt. Things would be a lot simpler if both these options were to be found in ceph.conf when they deviate from the default. That's almost what 
<pre>
        cryptsetup_parameters_str = get_conf(
            cluster=args.cluster,
            variable='osd_cryptsetup_parameters',
            )
        if cryptsetup_parameters_str is None:
            cryptsetup_parameters = []
        else:
            cryptsetup_parameters = shlex.split(cryptsetup_parameters_str)
</pre>
does in prepare_dev only it would be used for key-size and dmcrypt-key-dir. I suppose this what @ddiss refers to as "hardcoded path". 

That being said, since the auto-activation of dmcrypt is currently broken anyway, I don't think there is a need to hold up this pull request until this is fixed. Maybe aiming for something that autoactivate only when the default values are used is good enough for now ?

**Comment by dachary** [comment]
https://github.com/ceph/ceph/commits/wip-user looks good, I'll run the ceph-disk suite on it.

**Comment by dachary** [comment]
I'm seeing problems on CentOS 7
<pre>
[root@target186252 ceph-disk]# ceph-disk list
/dev/vda :
 /dev/vda1 other, xfs, mounted on /
/dev/vdb :
 /dev/vdb2 ceph journal, for /dev/vdb1
 /dev/vdb1 ceph data, prepared, cluster ceph, journal /dev/vdb2
/dev/vdc other, unknown
/dev/vdd other, unknown
[root@target186252 ceph-disk]# blkid -p -o udev /dev/vdb2
[root@target186252 ceph-disk]# 
</pre>
and in /var/log/messages
<pre>
Sep  1 17:50:44 target186252 ceph-disk: Traceback (most recent call last):
Sep  1 17:50:44 target186252 ceph-disk: File "/usr/sbin/ceph-disk", line 3575, in <module>
Sep  1 17:50:44 target186252 ceph-disk: main(sys.argv[1:])
Sep  1 17:50:44 target186252 ceph-disk: File "/usr/sbin/ceph-disk", line 3532, in main
Sep  1 17:50:44 target186252 ceph-disk: args.func(args)
Sep  1 17:50:44 target186252 ceph-disk: File "/usr/sbin/ceph-disk", line 3024, in main_trigger
Sep  1 17:50:44 target186252 ceph-disk: raise Error('no ID_PART_ENTRY_TYPE for %s' % args.dev)
Sep  1 17:50:44 target186252 ceph-disk: __main__.Error: Error: no ID_PART_ENTRY_TYPE for /dev/vdb2
Sep  1 17:50:44 target186252 systemd: ceph-disk@-dev-vdb2.service: main process exited, code=exited, status=1/FAILURE
Sep  1 17:50:44 target186252 systemd: Failed to start Ceph disk activation: /dev/vdb2.
Sep  1 17:50:44 target186252 systemd: Unit ceph-disk@-dev-vdb2.service entered failed state.
</pre>
which may or may not be related to the fact that I have a mixture of infernalis and wip-user manually installed to perform my tests. I'll figure it out today.

**Comment by dachary** [comment]
https://github.com/ceph/ceph/commits/wip-user is good to merge. Does not pass CentOS 7 tests (fails to activate because blkid fails on /dev/vdb2) but having these commits in infernalis is an improvement over the current situation anyway.

<pre>
Reviewed-by: Loic Dachary &lt;ldachary@redhat.com>
</pre>

**Pull request closed without merging**